### PR TITLE
Persistence plugin

### PR DESCRIPTION
**Pull request by JaurbanRH**
Change tests so make speedrun is compatible with pytest persistence plugin and add plugin dependency to Pipfile.
#179 needs to be merged first

### CODE REVIEW

File: testsuite/tests/apicast/policy/url_rewriting/test_url_rewriting_policy_append.py [hunk 1]
```diff
@@ -31,13 +31,16 @@ def service(service, private_base_url):
...
+
     echoed_request = EchoedRequest.create(api_client().get("/hello"))
     assert echoed_request.path == "/get"
 
     analytics = application.threescale_client.analytics
-    hits = analytics.list_by_service(application["service_id"], metric_name="get_metric")["total"]
+    hits_after = analytics.list_by_service(application["service_id"], metric_name="get_metric")["total"]
 
-    assert hits == 5
+    assert hits_before + 5 == hits_after
```
Comment (mdujava) [inline testsuite/tests/apicast/policy/url_rewriting/test_url_rewriting_policy_append.py:43]: values are swapped:
actual == expected

File: testsuite/tests/apicast/policy/url_rewriting/test_url_rewriting_policy_insert_above_apicast.py [hunk 2]
```diff
@@ -36,13 +36,16 @@ def service(service):
...
+
     echoed_request = EchoedRequest.create(api_client().get("/hello"))
     assert echoed_request.path == "/get"
 
     analytics = application.threescale_client.analytics
-    hits = analytics.list_by_service(application["service_id"], metric_name="get_metric")["total"]
+    hits_after = analytics.list_by_service(application["service_id"], metric_name="get_metric")["total"]
 
-    assert hits == 5
+    assert hits_before + 5 == hits_after
```
Comment (mdujava) [inline testsuite/tests/apicast/policy/url_rewriting/test_url_rewriting_policy_insert_above_apicast.py:48]: values are swapped:
actual == expected

**Review by mdujava — CHANGES_REQUESTED** [review]

File: testsuite/tests/apicast/auth/test_update_staging.py [hunk 3]
```diff
@@ -1,8 +1,10 @@
 """
 Rewrite: 3scale-amp-tests/spec/functional_specs/update_staging_spec.rb
 """
+import pytest
 
 
+@pytest.mark.sandbag  # incompatible with persistence plugin
```
Comment (mganisin) [inline testsuite/tests/apicast/auth/test_update_staging.py:7]: rm -rf

File: testsuite/tests/apicast/policy/batcher/test_batcher_policy_mapping_rules.py [hunk 4]
```diff
@@ -47,6 +47,7 @@ def client(api_client):
     return api_client(disable_retry_status_list={404})
 
 
+@pytest.mark.sandbag  # incompatible with persistence plugin
```
Comment (mganisin) [inline testsuite/tests/apicast/policy/batcher/test_batcher_policy_mapping_rules.py:50]: rm -rf

**Review by mdujava — DISMISSED** [review]

File: testsuite/tests/apicast/parameters/jaeger/test_jaeger_apicast_integration.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 from testsuite.utils import randomize
 from testsuite.capabilities import Capability
 
-CAPABILITIES = [Capability.JAEGER]
+CAPABILITIES = [Capability.JAEGER, Capability.CUSTOM_ENVIRONMENT]
```
Comment (pehala) [inline testsuite/tests/apicast/parameters/jaeger/test_jaeger_apicast_integration.py:13]: This is not a skip. (Not sure how it got through in the first place :D)

**Review by pehala — CHANGES_REQUESTED** [review]
I have found multiple failures in `system/` and `grafana/` (testsuite/tests/grafana/test_dashboards.py) directories.
Notably:
```
FAILED testsuite/tests/system/analytics/test_analytics.py::test_hits_service - assert 20 == (6 + 4)
FAILED testsuite/tests/system/analytics/test_backend_analytics.py::test_analytics - assert 10 == 5
FAILED testsuite/tests/system/mapping/test_mapping_rules_matching_order.py::test_mapping_rules_matching_order - assert 4 == 2
FAILED testsuite/tests/system/mapping/test_mapping_with_escape_sequence.py::test_mapping_with_escape_sequence - assert 2 == 1
```

**Comment by mganisin** [comment]
If I run with this I always see persistence in the list of loaded plugins. I'd like to rather not see it there unless explicitly requested. It's worth to experiment with `-p` pytest option

**Review by mganisin — CHANGES_REQUESTED** [review]
Probably I'd like to see something like `-p no:persistence` in default options in Makefile to **ensure** this plugin excluded unless explicitly requested. (e.g. `if not --(store|load) in flags then add -p no:persistence`)

**Review by mdujava — APPROVED** [review]

**Review by mganisin — DISMISSED** [review]

**Review by pehala — CHANGES_REQUESTED** [review]
A few tests are still failing, I believe a few of those that had their mark removed
```
FAILED testsuite/tests/apicast/test_hit_limit.py::test_retry_after - assert 200 == 429
FAILED testsuite/tests/apicast/apiap/test_backend_routing.py::test[empty_path-expect_ok1-expect_not_found1] - AssertionError: For path /anything/loud expected hits to be increased by 1
FAILED testsuite/tests/apicast/apiap/test_backend_routing.py::test[isolated_backends-expect_ok0-expect_not_found0] - AssertionError: For path /echo/anything/loud expected hits to be increased by 1
FAILED testsuite/tests/apicast/apiap/test_backend_routing.py::test[path_extension-expect_ok2-expect_not_found2] - AssertionError: For path /quotes/anything/test/anything/loud expected hits to be increased by 1
ERROR testsuite/tests/apicast/policy/tls/tls_upstream/tls_passthrough_connection_reuse/test_connection_reuse.py::test_connection_reuse - IndexError: list index out of range
ERROR testsuite/tests/system/test_unique_invoice_id.py::test_unique_invoice - threescale_api.errors.ApiClientError: Response(422 Unprocessable Entity): b'{"errors":{"users":["invalid"],"billing_address_name":[],"billing_address_ad...
ERROR testsuite/tests/system/test_unique_invoice_id.py::test_unique_invoice - threescale_api.errors.ApiClientError: Response(422 Unprocessable Entity): b'{"errors":{"base":["Cannot transition to cancelled"]}}'
ERROR testsuite/tests/system/mapping/test_mapping_rules_matching_order_multiple_backends.py::test[expect_ok0] - IndexError: list index out of range
```

**Review by pehala — APPROVED** [review]

**Review by mdujava — APPROVED** [review]

**Pull request merged**